### ISSUE TITLE
Query - How to get full list on Focus

### ISSUE THREAD

**Issue by raybooysen** [moroshko/react-autosuggest#382]
Consider the following use-case:

1)  The AutoSuggest has a value in the input
2) focus is lost and the user does something else
3) The user clicks on the AutoSuggest
4) The AutoSuggest opens
5) `_onSuggestionsFetchRequested` is called which, in my case, syncronously updates the suggestions based on the input.

`_onSuggestionsFetchRequested` is also called when you type into the input, as expected.

However, there is a subtle difference in how you want the suggestions to be rendered if you give focus to the AutoSuggest vs. typing into it.

If I give focus into an autosuggest, if it has a value in there, it would be nice to show, maybe all, or a bigger set of suggestions.  The user wants to choose a new value, so it would be good to show them a bigger list.  However, if a value is there already that scopes the suggestions down to the single value, they then have to delete text.

At the moment it looks like there is no hook for me to determine if it was the user giving the control focus or just typing into the auto suggest so that I can give more hints to the suggestion builder.

Any thoughts?

**Comment by raybooysen** [comment]
Having a look at the source code, this looks like it shouldn't be too hard:

https://github.com/moroshko/react-autosuggest/blob/master/src/Autosuggest.js#L469
On line 469, which is inside the onFocus event, the code calls onSuggestionFetchRequested, which could additionally pass some metadata about "why" the fetch has been requested.
```
if (shouldRender) {
            onSuggestionsFetchRequested({ value });
          }
```

**Comment by moroshko** [comment]
We could add `reason` to `onSuggestionsFetchRequested`, so the signature becomes:

```js
function onSuggestionsFetchRequested({ value, reason })
```

The possible values would be:
* `'suggestion-selected'`
* `'input-focused'`
* `'input-changed'`
* `'suggestions-revealed'`
* `'escape-pressed'`

What do you think?

Would you like to submit a PR?

**Comment by raybooysen** [comment]
Makes perfect sense.  I'll look at this today or over the weekend.

**Comment by raybooysen** [comment]
Code done, readme updated, but cannot push until tonight. :)

**Comment by raybooysen** [comment]
@moroshko I've just done a build and I have all the demo code rebuilt which has small whitespace changes, and my new reason added.  As part of the commit, do you want all these changes too?  Or just the raw code, test and readme?

**Comment by moroshko** [comment]
Just run `npm run prettier` - it should take care of the whitespace issues.

**Comment by raybooysen** [comment]
PR created and ready for review.

**Comment by raybooysen** [comment]
Thanks @moroshko .  Do you have a timeline of the next release?

**Comment by moroshko** [comment]
Hopefully, in the next couple of days.

**Comment by moroshko** [comment]
@raybooysen `reason` in available now is [v9.2.0](https://github.com/moroshko/react-autosuggest/releases/tag/v9.2.0).

**Comment by raybooysen** [comment]
Thanks!
